Added some basic library functions for the Int32

liampalmer43 opened the pull request:
@magnus-madsen 

- Added some basic library functions (and corresponding tests) for the Int32 type
    - minValue, maxValue, size, min, max, abs, dist, compare, signum
    - Bit operations: logicalRightShift, bitCount, rotateRight, rotateLeft

magnus-madsen commented on main/src/library/Int32.flix:
```diff
@@ -0,0 +1,88 @@
+            if (x % 2 != 0)
+                1 + bitCount(logicalRightShift(x, 1))
+            else
+                bitCount(logicalRightShift(x, 1))
```
You may be please to know about Flix' special switch expression:

```
def f(x: Int): Int = switch {
  case x < 0  => 1
  case x > 0  => 2
  case x == 0 => 3
}
```

Use it if you prefer.

magnus-madsen replied:
I'm just trying out this GitHub review feature. Let's see how it works.

liampalmer43 replied:
I've made use of Flix's switch statement, thanks Magnus.

olhotak commented on main/src/library/Int32.flix:
```diff
@@ -0,0 +1,88 @@
+    /**
+     * Returns the distance between `x` and `y`.
+     */
+    def dist(x: Int32, y: Int32): Int32 = abs(x - y)
```
Is this correct in case of overflow?

liampalmer43 replied:
I've updated dist to return -1 when the distance is too big for an Int32.  Thanks

olhotak commented on main/src/library/Int32.flix:
```diff
@@ -0,0 +1,88 @@
+namespace Int32 {
+
```
Are we implicitly assuming that Int32 represents signed two's-complement integers?

Or do we sometimes want unsigned integers as well?

mhyee replied:
The code generator already makes this assumption. It makes things a lot easier since we're targeting the JVM.

olhotak commented on main/src/library/Int32.flix:
```diff
@@ -0,0 +1,88 @@
+     */
+    def rotateRight(x: Int32, distance: Int32): Int32 =
+        (logicalRightShift(x, distance)) | (x << -distance)
+
```
I don't understand this. What does a shift by a negative distance do?

liampalmer43 replied:
Only the latter 5 bits are used to determine the shift distance when using >> and << (ie distance % 32).  So left-shifting by -distance is the same as left-shifting by (32 - distance).  Java uses this when implementing rotateRight and rotateLeft.

magnus-madsen commented:
At the same time, can I ask you to update the documentation:

https://github.com/flix/doc/blob/master/book/core.md

(You will need to clone that repository and make a separate pull request).

mhyee commented:
Just a reminder that if this is being merged in, `CONTRIBUTORS.md` needs to be updated (which means you agree to open-source your changes).

liampalmer43 commented:
- Addressed review comments
- Added some additional library functions (and corresponding tests) for the Int32 type
    - Bit operations: highestOneBitPosition, lowestOneBitPosition, highestOneBit, lowestOneBit, numberOfLeadingZeros, numberOfTrailingZeros, bit

magnus-madsen commented:
This is great work. You should also put in the copyright block to get some credit :)

```
/*
 * Copyright 2016 Liam Palmer
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
```

Looking forward to our discussion later today.

magnus-madsen commented on main/test/ca/uwaterloo/flix/library/TestInt32.scala:
```diff
@@ -0,0 +1,954 @@
+/**
+  * Created by liampalmer on 2016-09-18.
+  */
+class TestInt32 extends FunSuite {
```
Please remove this comment

magnus-madsen commented on main/test/ca/uwaterloo/flix/library/TestInt32.scala:
```diff
@@ -0,0 +1,954 @@
+  }
+
+  test("maxValue.01") {
+    val input =
```
If the input is a single line, I prefer to just write:

```val input = "f(42)"```

to keep it clean.

olhotak commented:
LGTM.
